drivers/serial/uart_nrfx_uarte.c

```diff
@@ -479,12 +478,20 @@ static int uarte_nrfx_tx(struct device *dev, const u8_t *buf, size_t len,
 		return -ENOTSUP;
 	}
 
-	if (data->async->tx_buf || data->async->tx_size) {
+	if (atomic_cas((atomic_t *) &data->async->tx_size,
```

Comment:
The atomic compare and set here is on async->tx_size , but the busy check in poll_out is on ansyc->tx_buf , so the pre-emption problem still exists.
1. Compare and set passes
2. Pre-empted
3. poll_out is called
4. Both threads think they are clear to use the hardware
I think you could just swap the check out to be on tx_buf, but I'm finding it hard to reason about the other ways tx_size is used.

Reply:
It has been taken into account, poll_out also has lock on async->tx_size and it will be blocked there in described scenario. Reason for using both tx_buf and tx_size for locks is that tx_buf is set only with async transfers, this way if poll_out is called from interrupt context then async transfer can be finished when trying to lock tx_buf, while tx_size is locked only in thread context, so if during poll_out its pre-empted and poll_out is called then we can still send from interrupt context. Only risk is that first byte sent using poll_out from interrupt context can be lost.
As using blocking functions in interrupt context is bad practice I assume that it would most likely occur only in hardfault handler, we don't want to lose that information so we cannot simply ignore poll out.
Current solution allows for using both async and polling API from different contexts without queueing output at cost of possible losing single byte when switching to interrupt context.

Reply:
Is that the only time bytes should be lost now? I have observed the first character of my test print disappearing with the setup described in my issue.

Reply:
Note that logger uses some multi byte encodings, depending on terminal settings it may seem like some characters are lost when logger thread is preempted in multi byte enconding. I noticed it when using ansi encoding in terminal, but looking at raw data I haven't seen it, therefore its not a UART issue.

Reply:
As you say, the raw bytes are there. Issue appears when we preempt after the first byte of the color code escape sequence. Thanks for the explanation, everything looks good to me then.